lib/std/priority_queue.zig

```diff
@@ -19,32 +19,34 @@ pub fn PriorityQueue(comptime T: type, comptime Context: type, comptime compareF
         const Self = @This();
 
         items: []T,
-        cap: usize,
-        allocator: Allocator,
+        /// The capacity of the queue. This may be read directly, but must not
+        /// be modified directly.
```

Comment:
* "the capacity of the queue" is 100% redundant with the name of the data structure and the field. Delete it.
* documentation should be descriptive, not prescriptive. Instead of telling the person what they should/shouldn't do, may or may not do, just document how it works.

```suggestion
        /// Tracks the allocated slice of memory when combined with `items.ptr`.
```

You can see how this comment is more useful, and the programmer can infer that they can read it directly, and that modifying it directly is in fact possible if they also modify `items` appropriately.

Missing documentation is acceptable. Documentation that violates the redundancy principle or descriptive principle is not. When in doubt, delete the documentation.

Reply:
I'm happy to change the comment and accept your suggestion. I do see how yours is more descriptive. 

I do however want to at least express my disagreement with this part:
> the programmer can infer that they can read it directly, and that modifying it directly is in fact possible if they also modify items appropriately.

I don't agree that the programmer either can, or should have to, infer which fields of a struct they are able to access directly. 

First, as someone who has had to navigate the standard library, I've found it a bit tedious that there's no standard way to get the capacity or current size of a container. I think for `ArrayList` I had to go read the tests to see that I'm supposed to access `items.len` directly to get the size of the container. While for another container (say, `PriorityDequeue`) you would have to do something different (use the `count()` function or `len` field). In the absence of standardization, documentation on a field to say that it's appropriate to access the field directly would be helpful so that I don't need to understand the implementation of the data structure just to get the size or capacity of it. I know that the standard library is still a work in progress and I think last I saw there's a plan to have a bit of a cleaning up/standardization later (is that true?) but I would hope that in the final incarnation, containers would have a consistent and documented way to fetch their size and capacity. And if that's via direct field access, I would hope that that would be documented. (My preference would be via a function with a name that isn't a noun - e.g. `getCapacity()` rather than `capacity()` which can conflict with variable names.)

Second, once back-compatibility starts to matter, it could be a big issue to have users accessing fields directly. Say a user is accessing the `items` field of `ArrayList` directly in order to obtain the current size of the list by doing `items.len`. If later on `items` were changed to store the allocated slice rather than the slice of current items, that would silently change the functionality of that users code, which would be a big problem if there were any promise of back-compatibility. (I know that isn't a change that's likely to be made, just an example :)) 

In other words, as soon as you say that users should directly access fields of a struct, you're effectively saying that fields become part of the API and now need to be considered as part of upgrades/deprecation/back-compatibility. This then limits your ability to swap out the internals of the data structure without impacting user code. If that is a choice that you're willing to make, then it seems like it would be a hard requirement to document which fields are expected to be accessed directly so that users can rely on that for stability and so maintainers know which fields need to undergo a deprecation process if they are removed or changed somehow. Otherwise, without this kind of documentation, I think all fields would need to be considered as public and unable to be removed/have their behavior changed without a deprecation process, since you're essentially saying they are all fair game to be accessed as the user sees fit.

Reply:
There are no private fields in Zig. All fields are public and part of the API, and all APIs must take this into account.

Reply:
Got it. I knew there was no support in the language for private fields but did not realize that the intent was also for all the fields to be conceptually part of the API. Thanks for explaining.